### DISCUSSION TITLE
BRouter long calculation time

### DISCUSSION THREAD

**Discussion by bios4** [devemux86/cruiser#379] [Q&A] [answered]
Since I started using cruiser I have been disappointed about the long time it takes to calculate routes.

To be able to show actual examples, i just tried to calculate two routes before writing this post...

Attempt 1 - from my current location to a biker meeting point, approx. 100 km away: Cruiser takes 17.81 seconds (then shows me three different routes); the Kurviger app calculates a route with the same start & destination in < 1 second (but without showing alternatives to be fair).

Attempt 2 - from my current location to a destination, approx. 260 km away: I can't say how long cruiser would need to do this because I aborted the calculation after two minutes (!) without a result; the Kurviger app calculates the route with the same start & destination in 1.64 seconds!

I'm currently running cruiser v3.0.36, routing is set to be done with BRouter (offline). All avoidances set to "off" except tertiary roads (80%), residential roads (40%) and avoid unpaved roads "on".

So my question is, if the calculation speed can be improved in any way?

Calculating routes with Cruiser and its current calculation performance is a PITA.
Thank god importing .kurviger-files is possible with Cruiser, so I am able to plan/create routes with the Kurviger app and navigate it with Cruiser later.

**Comment by devemux86** [comment]
@bios4 thanks for the report.

There is a big difference: BRouter is **offline**, all others are **online** services.
- Offline routing is done on the phone using the mobile (single) cpu.
- Online routing is done externally on fast (multiple) desktop servers.

For fast routing better use the an online routing service in Cruiser?
OpenRouteService or Valhalla also provide many routing options.

> three different routes

This means a slower calculation time **x 3**.
Better avoid alternative routes if want fast routing.

Also, using more waypoints makes routing much faster.

---

Please also write any experience in [BRouter repository](https://github.com/abrensch/brouter/), so that their developers know the opinions of users.

Routing speed was also discussed here:
- https://github.com/devemux86/cruiser/discussions/134#discussioncomment-7427507

cc @afischerdev @poutnikl

**Comment by WolfgangZ42** [comment]
The route calculation with BRouter (offline) is done completely offline - not on a server like Kurviger does it. So the "power" of your handheld influeces the time. And then you did three calculations at once and BRouter tries to differencate them totally. 
An example I did (btw: with your avoidances): 
<img width="200" src="https://github.com/devemux86/cruiser/assets/121806238/f5ebfe61-a054-4199-9675-afd1c7ad2eea"/>
I got this result in 40 seconds - fair enough for me.

**Comment by bios4** [comment]
Thanks for your responses...

I know that BRouter (offline) routing is done on my local device only (a Galaxy A52 btw, so not necessarily a very low performer).
But I did not expect that is THAT slow compared to other services.

> Also, using more waypoints makes routing much faster.

Adding waypoints would be something I most of the time do AFTER I have a basic route to my destination set up to fine-tune the route to my needs. So this does not help with the long calculation time of the initial route...

> I got this result in 40 seconds - fair enough for me.

Which routing service were you using here?
BRouter (online) or any other of the ones provided?

**Reply by devemux86** [reply]
@bios4 what happens with BRouter default settings?

- No alternative routes
- No avoidances
- Not change anything in the "Other" options screen

**Reply by WolfgangZ42** [reply]
> Which routing service were you using here?

Exactly the same as you mentioned: BRouter (offline) with your avoidance settings (plus avoid motorways!) - nothing else.

**Comment by bios4** [comment]
> @bios4 what happens with BRouter default settings?
> 
>     * No alternative routes
> 
>     * No avoidances
> 
>     * Not change anything in the "Other" options screen

With BRouter **(offline)** you mean?

Just gave it a try - the route from my attempt 1 in the initial post calculates in approx. 4 seconds, just by setting the "alternative routes" option back to 1.

The route from my attempt 2 does not calculate withing 2 minutes with the "alternative routes" set to 1, but it calculates the route in approx. 8 seconds when removing ALL options from BRouter (offline). Which (removing ALL options) is not a reasonable setting for me...

So I think I need to give it a try with one of the other services you mentioned - Valhalla seems to be the one which offers most similar options and avoidances to BRouter (offline).

**Reply by devemux86** [reply]
> the route from my attempt 1 in the initial post calculates in approx. 4 seconds, just by setting the "alternative routes" option back to 1.

:+1: 

> it calculates the route in approx. 8 seconds when removing ALL options from BRouter (offline).

Presumably for that route area these avoidances greatly _reduced the available road network_ for route finding.

A step-by-step process of gradually increasing the avoidances one by one may help to see which caused the problem.

**Comment by devemux86** [comment]
We can also use the Cruiser routing profile on the [BRouter website](https://brouter.de/brouter-web/) to plan routes with all options and avoidances.
Then export the planned routes and load them offline in Cruiser for navigation.

Cruiser routing profile and the instructions are here:
- https://github.com/devemux86/cruiser/discussions/272

**Comment by 0709wiwiwi** [comment]
**Offline** routing test. By phone Mi A2 Lite (A10).
Effective routing expected so no avoids where set.

- BRouter is fine for the shorter-distance routes.
- For long distance VERY slow and imo unsuitable.

**Offline Valhalla** router test. (By Guru map)
Generates up to 3 alternative designs.

Result. (Car)
(TR) Kusadasi > (TR) Ipsala. 3 sec 559 km.
(TR) Ipsala > (GR) Igoumenitsa. 3 sec 664 km.
(I) Bari > (I) Sona Verona. 3 sec 816 km.
(I) Sona Verona > (B) Genk. 4 sec 1044 km.

**Reply by devemux86** [reply]
They seem to use [contraction hierarchies](https://en.wikipedia.org/wiki/Contraction_hierarchies) like GraphHopper, OSRM (also fast routing).

What matters is if they have BRouter's flexibility to avoid literally every possible street type or OpenStreetMap tag.
Because many routing engines with such speed-up techniques have only a few avoidances, motorways, fords, etc.

---

So this is not a valid comparison, as already explained by @poutnikl here:
- https://github.com/devemux86/cruiser/discussions/134#discussioncomment-7427507

**Reply by 0709wiwiwi** [reply]
I report what I observed during long distance offline usage.
The facts and my observations, nothing more, nothing less

Practical.
The friend with TR & EU data with apple maps was the leader.
I so had sufficient time to do some practical tests.
Operation without data roaming in TR nor in CH.
All hotels and the ferry had already been booked.
Preparing the next day's journey by the hotels WiFi.
In TR we deliberately had deviated from the planned route.
The very slow BRouter calculation makes it unusable.
Absolutely not to be recommended as an offline solution.
None of the fellow travelers where convinced by the performance.
Anyway the offline experience for the hiking trajects was positive.

Note.
_The experimental Valhalla **online** routing appears to be an incomplete experimental setup. 
Valhalla even routes over private paths and the multiple settings react rather unpredictable. 
Maybe the Guru maps custom implementation is optimalised but I did not test that Guru mode.
As the BRouter solution already does offer nicely wat is necessary for such slow traffic routes._

**Reply by WolfgangZ42** [reply]
> The very slow BRouter calculation makes it unusable.
Absolutely not to be recommended as an offline solution.

I absolutely cannot understand this. All my recalculations are performed in seconds. However, I also set waypoints and do not calculate a distance of ~600 km (as in the examples above). 

Here are two recalculations:

<img width="300" src="https://github.com/devemux86/cruiser/assets/121806238/a11f44f6-eb96-49f9-8e16-d4b6315844ac"/>  <img width="300" src="https://github.com/devemux86/cruiser/assets/121806238/652a2d8a-1287-4323-817f-ea82887071ee"/>
This took less than a second

<img width="300" src="https://github.com/devemux86/cruiser/assets/121806238/9dda9ba8-9378-4fce-a4a7-87e4cf1cbfce"/>  <img width="300" src="https://github.com/devemux86/cruiser/assets/121806238/99bc4d10-c790-49b1-aedd-9144f167bd2f"/>
And this was ready in ~3 seconds 

If you want to see it live: [Ex.zip](https://github.com/devemux86/cruiser/files/13400784/Ex.zip)

**Reply by Leeb64** [reply]
When using the offline brouter you definitely have to think a bit.
Waypoints should be a maximum of 100km apart.
This shortens the calculation time and ensures that the planned route can be found again at some point in the event of a diversion.
By the way...
I wouldn't like to let a route planner like Kurviger plan a 500km tour without optimizing and correcting the whole thing.

**Reply by 0709wiwiwi** [reply]
> I absolutely cannot understand this. 

Surely I reported LONG DISTANCE TEST ! Efficient fast or shortest trajects routing !
I used pretty much the same method as described below. (Passing Dardanelles by the impressive bridge was expected.) 

> Adding waypoints would be something I most of the time do AFTER I have a basic route to my destination set up to fine-tune the route to my needs. So this does not help with the long calculation time of the initial route...

Is this successfully possible using ONLY Brouter offline data for fairly long routes? Test result ? = No. 
My expectation was the BRouter is probably unsuitable for this job and this has been confirmed. 
The test was instructive and a comparison was made with the fast Valhalla offline routing performance.
Experience has been reported here and so this need not cause any further difficult discussions.  Facts.

**Comment by bios4** [comment]
BTW - I'm late to the party but thank you @devemux86 for implementing the internal BRouter (offline) option in the current Beta! 👍 
It is a significant improvement in calculating my (test) routes from the initial post ^^...!

**Comment by devemux86** [comment, marked as answer]
BRouter calculation time was greatly improved in [Cruiser 3.0.37](https://play.google.com/store/apps/details?id=gr.talent.cruiser).
